core/kotlinx-coroutines-test/src/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,294 @@
+     * time and execute coroutines scheduled in the future use, one of [advanceTimeBy],
+     * or [advanceUntilIdle].
+     */
+    @ExperimentalCoroutinesApi
+    fun resumeDispatcher()
+}
+
+/**
+ * Thrown when a test has completed and there are tasks that are not completed or cancelled.
+ */
+@ExperimentalCoroutinesApi
+class UncompletedCoroutinesError(message: String, cause: Throwable? = null): AssertionError(message, cause)
+
+/**
+ * [CoroutineDispatcher] that can be used in tests for both immediate and lazy execution of coroutines.
+ *
+ * By default, [TestCoroutineDispatcher] will be immediate. That means any tasks scheduled to be run without delay will
+ * be immediately executed. If they were scheduled with a delay, the virtual clock-time must be advanced via one of the
+ * methods on [DelayController]
+ *
+ * When swiched to lazy execution any coroutines started via [launch] or [async] will
+ * not execute until a call to [DelayController.runCurrent] or the virtual clock-time has been advanced via one of the
+ * methods on [DelayController].
+ *
+ * @see DelayController
+ */
+@ExperimentalCoroutinesApi
+class TestCoroutineDispatcher:
+        CoroutineDispatcher(),
+        Delay,
+        DelayController {
+
+    private var dispatchImmediately = true
+        set(value) {
+            field = value
+            if (value) {
+                // there may already be tasks from setup code we need to run
+                advanceUntilIdle()
+            }
+        }
+
+    // The ordered queue for the runnable tasks.
+    private val queue = ThreadSafeHeap<TimedRunnable>()
+
+    // The per-scheduler global order counter.
+    private var counter = AtomicLong(0)
+
+    // Storing time in nanoseconds internally.
+    private var time = AtomicLong(0)
+
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        if (dispatchImmediately) {
+            block.run()
+        } else {
+            post(block)
+        }
+    }
+
+    override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) {
+        postDelayed(CancellableContinuationRunnable(continuation) { resumeUndispatched(Unit) }, timeMillis)
+    }
+
+    override fun invokeOnTimeout(timeMillis: Long, block: Runnable): DisposableHandle {
+        val node = postDelayed(block, timeMillis)
+        return object : DisposableHandle {
+            override fun dispose() {
+                queue.remove(node)
+            }
+        }
+    }
+
+    override fun toString(): String {
+        return "TestCoroutineDispatcher[currentTime=${time}ms, queued=${queue.size}]"
+    }
+
+    private fun post(block: Runnable) =
+            queue.addLast(TimedRunnable(block, counter.getAndIncrement()))
+
+    private fun postDelayed(block: Runnable, delayTime: Long) =
+            TimedRunnable(block, counter.getAndIncrement(), time.get() + delayTime)
```

Comment:
When writing the old `TestCoroutineContext`, I forgot to add tests for `Long` overflows in the dispatcher.

If the code-under-test tries to simulate an always-suspending coroutine by calling `delay(Long.MAX_VALUE)`, the TimedRunnable would be scheduled to run immediately, since `time.get() + delayTime` would result in a negative value. 

Be sure to clamp the sum `time.get() + delayTime` against Long.MAX_VALUE so that this sum would not accidentally become a negative value.

Reply:
I took care of it.